source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+          source: *stream0
+        expectedResult: { $$type: "objectId" }
+        saveResultAsEntity: &objectid0 "objectid0"
+      - object: *filesCollection
+        name: findOne
+        arguments:
+          sort: { uploadDate: -1 }
+        expectedResult:
+          _id: { $$matchesEntity: *objectid0 }
+
+
+$$matchesHexBytes
+`````````````````
+
+Syntax, where ``hexBytes`` is an even number of hexademical characters
+(case-insensitive)::
+
+    { $$matchesHexBytes: <hexBytes> }
+
+This operator can be used anywhere a matched value is expected (including an
+`expectedResult <operation_expectedResult_>`_) and the actual value is a stream
+as defined in the `GridFS <../gridfs/gridfs-spec.rst>`__ spec. The test runner
+MUST convert the string to a byte sequence and compare it with the full contents
+of the stream. The test runner MUST raise an error if the string is malformed.
+
+This operator is primarily used to assert the contents of stream returned by
+`openDownloadStream`_ and `openDownloadStreamByName`_.
+
+
+$$unsetOrMatches
+````````````````
+
+Syntax::
+
+    { $$unsetOrMatches: <anything> }
+
+This operator can be used anywhere a matched value is expected (including an
+`expectedResult <operation_expectedResult_>`_). The test runner MUST assert that
+actual value either does not exist or matches the expected value. Matching the
+expected value should use the standard rules in `Evaluating Matches`_, which
+means that it may contain special operators.
+
+This operator is primarily used to assert driver-optional fields from the CRUD
+spec (e.g. ``insertedId`` for InsertOneResult, ``writeResult`` for
+BulkWriteException).
+
+An example of this operator used for a result's field follows::
+
+    expectedResult:
+      insertedId: { $$unsetOrMatches: 2 }
+
+An example of this operator used for an entire result follows::
+
+    expectedError:
+      expectedResult:
+        $$unsetOrMatches:
+          deletedCount: 0
+          insertedCount: 2
+          matchedCount: 0
+          modifiedCount: 0
+          upsertedCount: 0
+          upsertedIds: { }
+
+
+$$sessionLsid
+`````````````
+
+Syntax::
+
+    { $$sessionLsid: <string> }
+
+This operation is used for matching any value with the logical session ID of a
+`session entity <entity_session_>`_. The value will refer to a unique name of a
+session entity. The YAML file SHOULD use an `alias node`_ for a session entity's
+``id`` field (e.g. ``session: *session0``).
+
+An example of this operator follows::
+
+    command:
+      ping: 1
+      lsid: { $$sessionLsid: *session0 }
+
+
+Test Runner Implementation
+--------------------------
+
+The sections below describe instructions for instantiating the test runner,
+loading each test file, and executing each test within a test file. Test runners
+SHOULD NOT share state created by processing a test file with the processing of
+subsequent test files, and likewise for tests within a test file.
+
+
+Initializing the Test Runner
```

Comment:
In order to minimize the implementation logic and open up some of the functionality could we consider an `initialOperations` top-level key over `initialData` where `initialOperations` can be any operation? This could way more useful without any extra logic on the runners. Could be useful for setting up indexes.

Reply:
Most test runners in node have hooks, `before` aka `beforeAll`, `beforeEach`, `after`  aka `afterAll`, `afterEach`.  

The term `initial` could have two meanings, one being at `before`, or `beforeEach`, given the spec it seems to be `beforeEach`. I'm not sure we'd ever want a `before` or `after` operations hook, but it may be useful, we could leave a `before` implementation out of the picture for now, but leave the window open for it in the future.

A more explicit approach could look like this:

````
{
    operationHooks: {
        beforeEach: [
            {<insert data>} ,
        ],
        afterEach: [
            {<drop collection>} ,
        ]
    }
}
```

Or

````
{
    operationHooks: {
        beforeEach: [
            {<drop collection>} ,
            {<insert data>} ,
        ],
    }
}
```

This would elevate the manual requirement of requiring the runner to drop the collection. We may never want the data to persist between tests, so this may not be as useful for that case.

Reply:
This change also has the potential to deprecate `createEntities`, if any `executable` (entity / operation) result can be storied as an entity, the creation of initial entities (client, database, collection) could be run in `before`, and contain `entityId`s, vs a `insertOne` or `insertMany` wound't need to have an `entityId`. Again similar wheelhouse of trying to merge the idea of an entity and operation.

Reply:
Interestingly we could have no `createEntities` and run tests against client / database / collection if they are treated the same as any operation.

Reply:
Still riffing on this:

```json
{
  "description": "Example",
  "schemaVersion": "1.0",
  "executeBefore": [
    {
      "reference": "client",
      "entityId": "client-ref",
      "parameters": {
        "uriOptions": {
          "replicaSet": "ab"
        }
      }
    },
    {
      "reference": "database",
      "entityId": "db-ref",
      "parameters": {
        "clientEntityId": "client-ref",
        "databaseName": "test"
      }
    },
    {
      "reference": "collection",
      "entityId": "col-ref",
      "parameters": {
        "databaseEntityId": "db-ref",
        "databaseName": "test"
      }
    }
  ],
  "executeBeforeTest": [
    {
      "reference": "insertMany",
      "entityId": "initial-data",
      "parameters": {
        "databaseEntityId": "db-ref",
        "documents": [
          { "foo": true },
          { "bar": true },
          { "baz": true }
        ]
      }
    }
  ],
  "executeAfterTest": [
    {
      "reference": "dropCollection",
      "parameters": {
        "databaseEntityId": "db-ref",
        "collectionName": "test"
      }
    }
  ],
  "tests": {
    "description": "Tests findOne with collection data.",
    "execute": [
      {
        "reference": "findOne",
        "parameters": {
          "collectionEntityId": "col-ref",
          "filter": { "bar": true }
        },
        "expectSuccess": {
          "result": {
            "bar": {
              "$$exists": true
            }
          }
        }
      },
      {
        "reference": "findOne",
        "parameters": {
          "collectionEntityId": "col-ref",
          "query": { "howdy": true }
        },
        "expectFailure": {
          "$$exists": true
        }
      },
      {
        "reference": "find",
        "parameters": {
          "collectionEntityId": "col-ref"
        },
        "expectSuccess": {
          "result": {
            "documents": {
              "$$matchesEntity": "initial-data"
            }
          }
        }
      }
    ]
  }
}
```

Reply:
Looking at the last example, it looks like `reference` is the unique identifier of an executable item since the contextual object for the method is placed under `parameters`. API methods and custom testRunner operations defined in the spec shouldn't share a common namespace, so I would suggest keeping the operation name and object as top-level fields.

In an earlier comment, I explained that `session` is probably the only entity for which the creation method _is_ specified (`MongoClient.startSession`). For other entities, I suppose we could create specific testRunner operations to create them. For example:

```yaml
operations:
  - object: testRunner
    name: createClient
    parameters:
      observeEvents: [ "commandStartedEvent" ]
      useMultipleMongoses: false
    saveResultAsEntity: &client0 "client0"
  - object: testRunner
    name: createDatabase
    parameters:
      client: *client0
      databaseName: "test"
    saveResultAsEntity: &database0 "database0"
```    

A potential caveat is that tests would need to conditionally create separate entities if they could not use an entity under test for a particular pre- or post-test operation. Additionally, tests would need to manually specify a "majority" write concern for some spec tests, which itself would likely require separate entities to be created since per-operation write concerns are no universally specified for, and supported by, all drivers.

Reply:
Originally my thinking here is that every standardized (driver-wide) method is almost a unique `key`, and drivers implement a mapping to that key, and it doesn't have to match exactly, it can be mapped my the driver. I see the approach we have is much more explicit when it comes to what is executing. Another way to think about that is that every executable is a unique identifier like `testRunner`, `collection.insertOne`, `client.create`, etc and it's up to the driver to map that executable operation. I see we have it differently here where we try to minimize those mappings and the `testRunner` takes care of unique operations. It's just a difference in approaches.

Reply:
Noted. I see how having unique identifiers with `<entity_type>.<method>` could be clear and allow us to enforce unique names, but if test authors are following the guidance in the design then each entity will be named after its type (e.g. `client0`, `session1`) and I think that would be just as readable. It also simplifies test runner code that would need to parse the identifier and any edge case where the entity type we dynamically fetch at runtime doesn't match the `<entity_type>` from `name`.

Thanks for the dialog on this and fleshing out examples to present your suggestions.